File: c/RNG31Core/AbstractRNG31Core.c
#include "AbstractRNG31Core.h"

void rng31Core_initDefault(AbstractRNG31Core *rng, AbstractRNG31CoreInit initPtr, AbstractRNG31CoreNext nextPtr)
{
    rng31Core_init(rng, initPtr, nextPtr, 0);
}

void rng31Core_init(AbstractRNG31Core *rng, AbstractRNG31CoreInit initPtr, AbstractRNG31CoreNext nextPtr, int32_t seed)
{
    rng->initPtr = initPtr;
    rng->nextPtr = nextPtr;
    rng31Core_reseed(rng, seed);
}

void rng31Core_reseedDefault(AbstractRNG31Core *rng)
{
    rng31Core_reseed(rng, 0);
}

void rng31Core_reseed(AbstractRNG31Core *rng, int32_t seed)
{
    if(seed == 0)
        seed = (int)time(NULL);
    rng->initialSeed = seed;
    rng->initPtr(rng);
}

int32_t rng31Core_next(AbstractRNG31Core *rng)
{
    int32_t result = rng->nextPtr(rng);
    assert(0 <= result && result <= ABSTRACT_RNG31CORE_MAX);
    return result;
}

int32_t rng31Core_initialSeed(AbstractRNG31Core *rng)
{
    return rng->initialSeed;
}

File: py/RNG31Core/AbstractRNG31Core.py
import time


class AbstractRNG31Core:
    """
    Abstract base class for generating random numbers in [0, 2^31 - 1].
    - Implement _initialize() to set up the initial state. This is called by reseed().
    - Implement _next(). This must generate a number in the range above.
    """

    MAX = 0x7FFFFFFF

    def __init__(self, seed=0):
        self.__initialSeed = 0

        self.reseed(seed)

    def _initialize(self):
        raise NotImplementedError

    def _next(self):
        raise NotImplementedError

    def next(self):
        result = self._next()
        assert(0 <= result and result <= AbstractRNG31Core.MAX)
        return result

    def reseed(self, seed):
        if seed == 0:
            seed = int(time.time())
        self.__initialSeed = seed
        self._initialize()

    def initialSeed(self):
        return self.__initialSeed
